File: android/src/main/java/com/easyhrworld/capacitor_background_geolocation/HeadlessHttpPoster.java
package com.easyhrworld.capacitor_background_geolocation;

import android.content.Context;
import android.content.SharedPreferences;
import com.getcapacitor.Logger;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import org.json.JSONArray;
import org.json.JSONObject;

public class HeadlessHttpPoster {

    private static final String PREFS_NAME = "bg_geo_prefs";
    private final Context context;

    public HeadlessHttpPoster(Context context) {
        this.context = context;
    }

    public void postBatch(LocationBuffer locationBuffer) {
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        String serverUrl = prefs.getString("headless_server_url", "");
        String authToken = prefs.getString("headless_auth_token", "");
        String employeeId = prefs.getString("headless_employee_id", "");
        String tenantId = prefs.getString("headless_tenant_id", "");
        int batchSize = prefs.getInt("headless_batch_size", 20);

        if (serverUrl.isEmpty() || authToken.isEmpty()) {
            Logger.debug("Headless posting not configured, skipping");
            return;
        }

        JSONArray batch = locationBuffer.getUnsyncedBatchAsJson(batchSize);
        if (batch.length() == 0) {
            return;
        }

        HttpURLConnection conn = null;
        try {
            JSONObject payload = new JSONObject();
            payload.put("employeeId", employeeId);
            payload.put("tenantId", tenantId);

            // Build locations array without the internal 'id' field
            JSONArray locations = new JSONArray();
            for (int i = 0; i < batch.length(); i++) {
                JSONObject src = batch.getJSONObject(i);
                JSONObject loc = new JSONObject();
                loc.put("lat", src.getDouble("lat"));
                loc.put("lng", src.getDouble("lng"));
                loc.put("accuracy", src.getDouble("accuracy"));
                loc.put("speed", src.getDouble("speed"));
                loc.put("bearing", src.getDouble("bearing"));
                loc.put("altitude", src.getDouble("altitude"));
                loc.put("timestamp", src.getLong("timestamp"));
                locations.put(loc);
            }
            payload.put("locations", locations);

            byte[] body = payload.toString().getBytes(StandardCharsets.UTF_8);

            conn = (HttpURLConnection) new URL(serverUrl).openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Authorization", "Bearer " + authToken);
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setRequestProperty("Content-Length", String.valueOf(body.length));
            conn.setConnectTimeout(30000);
            conn.setReadTimeout(30000);
            conn.setDoOutput(true);

            OutputStream os = conn.getOutputStream();
            os.write(body);
            os.flush();
            os.close();

            int responseCode = conn.getResponseCode();
            if (responseCode >= 200 && responseCode < 300) {
                Logger.info("Headless post successful: " + batch.length() + " locations");
                locationBuffer.markSynced(batch);
                locationBuffer.deleteSynced();
            } else {
                Logger.error("Headless post failed with HTTP " + responseCode);
            }
        } catch (Exception e) {
            Logger.error("Headless HTTP post error", e);
        } finally {
            if (conn != null) {
                conn.disconnect();
            }
        }
    }
}

File: ios/Sources/CapgoBackgroundGeolocationPlugin/HeadlessHttpPoster.swift
import Foundation

/// Posts buffered location batches to a server endpoint using native URLSession.
/// Mirrors the Android HeadlessHttpPoster.java implementation.
class HeadlessHttpPoster {

    private static let prefsPrefix = "bg_geo_"

    /// Post a batch of unsynced locations to the configured server.
    /// Reads configuration from UserDefaults.
    func postBatch(_ locationBuffer: LocationBuffer) {
        let defaults = UserDefaults.standard
        let serverUrl = defaults.string(forKey: "\(Self.prefsPrefix)server_url") ?? ""
        let authToken = defaults.string(forKey: "\(Self.prefsPrefix)auth_token") ?? ""
        let employeeId = defaults.string(forKey: "\(Self.prefsPrefix)employee_id") ?? ""
        let tenantId = defaults.string(forKey: "\(Self.prefsPrefix)tenant_id") ?? ""
        let batchSize = defaults.integer(forKey: "\(Self.prefsPrefix)batch_size")
        let effectiveBatchSize = batchSize > 0 ? batchSize : 20

        guard !serverUrl.isEmpty, !authToken.isEmpty else {
            print("[BackgroundGeolocation] Headless posting not configured, skipping")
            return
        }

        guard let url = URL(string: serverUrl) else {
            print("[BackgroundGeolocation] Invalid server URL: \(serverUrl)")
            return
        }

        let batch = locationBuffer.getUnsyncedBatch(effectiveBatchSize)
        guard !batch.isEmpty else { return }

        // Build payload matching Android format
        var locations: [[String: Any]] = []
        var ids: [Int64] = []

        for row in batch {
            if let id = row["id"] as? Int64 {
                ids.append(id)
            }
            var loc: [String: Any] = [:]
            loc["lat"] = row["lat"]
            loc["lng"] = row["lng"]
            loc["accuracy"] = row["accuracy"]
            loc["speed"] = row["speed"]
            loc["bearing"] = row["bearing"]
            loc["altitude"] = row["altitude"]
            loc["timestamp"] = row["timestamp"]
            locations.append(loc)
        }

        let payload: [String: Any] = [
            "employeeId": employeeId,
            "tenantId": tenantId,
            "locations": locations,
        ]

        guard let body = try? JSONSerialization.data(withJSONObject: payload) else {
            print("[BackgroundGeolocation] Failed to serialize payload")
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 30
        request.httpBody = body

        // Synchronous post on background thread (called from timer)
        let semaphore = DispatchSemaphore(value: 0)
        var success = false

        let task = URLSession.shared.dataTask(with: request) { _, response, error in
            defer { semaphore.signal() }

            if let error = error {
                print("[BackgroundGeolocation] Headless HTTP post error: \(error.localizedDescription)")
                return
            }

            if let httpResponse = response as? HTTPURLResponse,
               (200..<300).contains(httpResponse.statusCode) {
                print("[BackgroundGeolocation] Headless post successful: \(batch.count) locations")
                success = true
            } else {
                let code = (response as? HTTPURLResponse)?.statusCode ?? -1
                print("[BackgroundGeolocation] Headless post failed with HTTP \(code)")
            }
        }
        task.resume()
        semaphore.wait()

        if success {
            locationBuffer.markSynced(ids)
            locationBuffer.deleteSynced()
        }
    }
}
